suitesparse: 4.4.4 -> 5.3.0

jluttine opened the pull request:
###### Motivation for this change

~~**This is work in progress, do not merge.**~~

**EDIT: Ready for review and merge now.**

Upgrading SuiteSparse to 5.3.0. ~~However, the build failed and I was a bit lost with all the complexity of the package. I thought I'd put it here if someone wants to help and/or that it's public so there wouldn't be overlapping efforts done.~~

~~[Error log](https://github.com/NixOS/nixpkgs/files/1739323/suitesparse.log)~~

~~cc: @ttuegel Perhaps you, as a maintainer, know better how to upgrade this package?~~

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

jluttine committed:
suitesparse: 4.4.4 -> 5.3.0

jluttine commented:
I modified the derivation quite heavily. Someone should check that it still makes sense. Anyway, the build is almost successful now. At the very end, some phase after installation I suppose, I get the following error:

```
/nix/store/4gc1xjghbqfwc498cidj53dk5ynwr0zz-suitesparse-5.1.2/share/doc/suitesparse-5.1.2/cholmod/getproto: interpreter directive changed from "/bin/sh" to "/nix/store/i0ay05pqkbnvpfijm52mmlrp6kmkl80c-bash-4.4-p12/bin/sh"
/nix/store/4gc1xjghbqfwc498cidj53dk5ynwr0zz-suitesparse-5.1.2/share/doc/suitesparse-5.1.2/cholmod/getmproto: interpreter directive changed from "/bin/sh" to "/nix/store/i0ay05pqkbnvpfijm52mmlrp6kmkl80c-bash-4.4-p12/bin/sh"
checking for references to /tmp/nix-build-suitesparse-5.1.2.drv-0 in /nix/store/4gc1xjghbqfwc498cidj53dk5ynwr0zz-suitesparse-5.1.2...
RPATH of binary /nix/store/4gc1xjghbqfwc498cidj53dk5ynwr0zz-suitesparse-5.1.2/lib/libspqr.so.2.0.8 contains a forbidden reference to /tmp/nix-build-suitesparse-5.1.2.drv-0
builder for ‘/nix/store/1wzlj7r2ppmqg6biibjsny9ijj6jdcwg-suitesparse-5.1.2.drv’ failed with exit code 1
```

Not yet sure how to fix it.

jluttine commented:
This built now on my machine. I think someone should review this as I'm not too confident about my changes.

jluttine commented:
Hmm.. nox-review failing.

jluttine commented:
@GrahamcOfBorg build julia

jluttine commented:
@GrahamcOfBorg build suitesparse

jluttine commented:
I changed the updated version from 5.2.0 to 5.3.0. Build succeeds, but nox review fails:

```
error: build of '/nix/store/f8p1b61nwdamwyx9ff897d9ps4i4p8ly-sage-wrapper-8.3.drv', '/nix/store/rjfqbak8b89123zbzwir1ich9pc1yfyb-sage-wrapper-8.3.drv', '/nix/store/s0cwmcvix092b8fh518ck18zv479mnxg-julia-0.6.4.drv' failed
The invocation of "nix-build -A suitesparse -A sageWithDoc -A python36Packages.cvxopt -A python27Packages.cvxopt -A sage -A julia -A suitesparse_4_4 /etc/nixpkgs" failed
```

jluttine commented:
Fixed the symbol loading issue, but still nox review failing.
```
error: build of '/nix/store/9rpg8fnlw5gd3c9m9lrdw635xj0jznw3-julia-0.6.4.drv', '/nix/store/ibmnf9qxqm1w4g510d3vzk11gm6z8znx-sage-wrapper-8.3.drv', '/nix/store/xaj5bqv6rhm2jq4vm834bx0kqvdpcrm1-sage-wrapper-8.3.drv' failed
The invocation of "nix-build -A suitesparse -A sageWithDoc -A python27Packages.cvxopt -A python36Packages.cvxopt -A sage -A julia -A suitesparse_4_4 /etc/nixpkgs" failed
```

jluttine commented:
I tried building `julia` without this pull request on nixos-unstable but it didn't build. I got the same error. So, I suppose that package is broken despite this change.

ttuegel commented on pkgs/development/libraries/science/math/suitesparse/default.nix:
```diff
@@ -47,48 +50,45 @@ stdenv.mkDerivation {
 
-  postInstall = ''
-    # Build and install shared library
+  buildPhase = ''
```
Please invoke the pre- and post-phase hooks in a custom phase, i.e.
```.nix
{
buildPhase = ''
  runHook preBuild

  # ...

  runHook postBuild
'';
}
```
Other packages' setup hooks can add steps to the pre- and post-phase hooks which will not run otherwise. The same comment applies to `installPhase` below.

jluttine replied:
Thanks, done.

jluttine commented on pkgs/development/libraries/science/math/suitesparse/default.nix:
```diff
@@ -1,25 +1,28 @@
+    sha256 = "0gcn1xj3z87wpp26gxn11k8073bxv6jswfd8jmddlm64v09rgrlh";
   };
 
+  dontUseCmakeConfigure = true;
```
Can't say I understand much about these, but if `cmake` is not a build input, it fails to build GraphBLAS.

jluttine commented:
Hmm.. I noticed some things aren't working. I'll check some BLAS-related build options tomorrow, could be related to those.

On October 8, 2018 5:02:01 PM GMT+03:00, Thomas Tuegel <notifications@github.com> wrote:
>ttuegel commented on this pull request.
>
>
>
>>    };
> 
>+  dontUseCmakeConfigure = true;
>
>What is the dependency on `cmake` for if not calling `cmake`?
>
>>    NIX_CFLAGS_COMPILE = stdenv.lib.optionalString stdenv.isDarwin "
>-DNTIMER";
> 
>-  postInstall = ''
>-    # Build and install shared library
>+  buildPhase = ''
>
>Please invoke the pre- and post-phase hooks in a custom phase, i.e.
>```.nix
>{
>buildPhase = ''
>  runHook preBuild
>
>  # ...
>
>  runHook postBuild
>'';
>}
>```
>Other packages' setup hooks can add steps to the pre- and post-phase
>hooks which will not run otherwise. The same comment applies to
>`installPhase` below.

jluttine commented:
Ok, that should be fixed now. For some reason, I had added `-llapack` build flag and `liblapack` as a build dependency. These weren't compatible with openblas, which contains its own lapack anyway, so I just removed lapack as it had been before. As far as I can tell, this should be working now. Any other comments, @ttuegel?

ttuegel commented:
Thanks!
